hal/src/thumbv7em/clock/v2.rs

```diff
@@ -1,98 +1,540 @@
-//#![warn(missing_docs)]
-
-//! # Clocking API v2
+//! # Version 2 of the `clock` module
 //!
-//! This module provides the set of abstractions allowing a user to manipulate
-//! an ATSAMD's clocking system in a safe and expressive manner.
+//! ## Overview
 //!
-//! Foundation of an API is a [`retrieve_clocks`] function that returns a tuple
-//! of instantiated clocking abstractions that are enabled on reset.
+//! This module provides a simple, ergonomic, and most of all **safe** API to
+//! create and manage the clock tree in ATSAMD5x and E5x devices. It uses
+//! [type-level programming techniques](crate::typelevel) to prevent users from
+//! creating invalid or unsound clocking configurations.
 //!
-//! To reconfigure a clocking tree, user can instantiate additional clocking
-//! components or reconfigure existing ones and connect them. Type-safe nature
-//! of the API prevents user from building an application with an unsound clock
-//! tree setup.
+//! <p style="background:rgba(255,181,77,0.16);padding:0.75em;">
+//! <strong> Note: </strong> Using a type-level API does place some limitations
+//! on how the clocks can be configured. The types must be checked at
+//! compile-time, which means the state of a given clock must also be known at
+//! compile-time. This is exceedingly reasonable for most clocking
+//! configurations, because most users set up their clocks once and never change
+//! them again. However, if you need to dynamically change the clocking
+//! configuration at run-time, you may find this API less ergonomic. A future,
+//! fully dynamic API has been discussed, but nothing has been developed so far.
+//! </p>
 //!
-//! Example:
-//! ```no_run
-//! # use atsamd_hal::{
-//! #     clock::v1,
-//! #     clock::v2::{
-//! #         por_state,
-//! #         apb::{ApbClk, ApbToken},
-//! #         gclkio::{GclkIn, GclkOut},
-//! #         xosc::{CrystalCurrent, Xosc},
-//! #         gclk::{Gclk, Gclk1Div},
-//! #         dpll::Dpll,
-//! #         pclk::Pclk,
-//! #         types::Trng,
-//! #     },
-//! #     gpio::Pins,
-//! #     pac::Peripherals,
-//! #     time::U32Ext,
-//! # };
+//! There are a few basic clock categories in the ATSAMD clock tree.
+//! Specifically, there are:
+//! - Root clocks derived from internal clocks, external clocks or crystal
+//!   oscillators:
+//!     - The internal, 48 MHz DFLL ([`dfll`])
+//!     - The internal, ultra-low power oscillator ([`osculp32k`])
+//!     - An external clock or crystal oscillator ([`GclkIn`], [`xosc`], [`xosc32k`])
+//! - Branch clocks that modify and distribute clocks:
+//!     - Generic clock generators ([`gclk`])
+//!     - Digital phase-locked loops ([`dpll`])
+//! - Leaf clocks that drive peripherals or external clock outputs:
+//!     - Peripheral channel clocks ([`pclk`])
+//!     - The real-time clock oscillator ([`rtcosc`])
+//!     - External clock ouputs ([`GclkOut`])
+//! - Bus clocks derived from the processor's master clock:
```

Comment:
master clock -> main clock to be consistent with the datasheet MCLK. The only master clocks I find there are in I2C, I2S and ethernet peripherals.

Reply:
Done